### ISSUE TITLE
Model Inferencing Issue for LSTM  Openvino IR model

### ISSUE THREAD

**Issue by CosmicHunter** [openvinotoolkit/openvino#1273]
I have created a sentiment classification LSTM model in pytorch , The model is working fine , Then I converted it to ONNX and checked the onnx model also using the onnx-runtime . Now I converted the onnx model to Openvino IR . But for different inputs the model is always producing the same output. I am not able to determine the problem . I am also attaching the output when I generate the model to IR , And also the code snippet used for inferencing . My Lstm model is of 2 layers , hidden dimension  = 256 ,  and sequences are padded with length 200 before feeding it to the model . So the input is of fixed length.

## Code Snippet for Generating IR
~~~
===> input _shape = (batch_size , sequence_len , embedding dimension )
! /opt/intel/openvino/deployment_tools/model_optimizer/mo_onnx.py --input_model   /content/lstm_model_2_test.onnx --data_type=FP32 --output_dir fp32
~~~

## Output 
![out1](https://user-images.githubusercontent.com/52317137/87081286-b47fe000-c246-11ea-90ba-4ca4fab2e8c9.PNG)

## Code Snippet Used For Inferencing
 The models forward function expects a sequence of length of 200 and hidden state and cell state of dimension (2,1,256) where is 2 is number of layers and 1 is the batch size  and 256 is the hidden dimension

~~~
    plugin = IEPlugin(target_device)  
    args = build_argparser().parse_args()
    model_xml = args.model
    model_bin = os.path.splitext(model_xml)[0] + ".bin"
    log.info("Loading network files:\n\t{}\n\t{}".format(model_xml, model_bin))
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    num_requests = 1
    net = IENetwork(model=model_xml, weights=model_bin)
    input_blob1 = next(iter(net.inputs))
    input_blob2 = next(iter(net.inputs))
    input_blob3 = next(iter(net.inputs))
    hiden_init_state = torch.zeros(2,1,256)
    hiden_init_state = hiden_init_state.to(device)     
    # inputs = args.input[0]   #for getting input sentence  
    count = 0
    result = device_model.infer({input_blob1: inputs,
                                 input_blob2 : hiden_init_state,
                                 input_blob3: hiden_init_state})
    
    request_id = 0
    status = device_model.requests[request_id].wait(-1)
    
    if status == 0:
        res = device_model.requests[request_id].outputs
    print(res['Squeeze_24'])
    print(res.keys())
~~~

## Output of Inferencing

![out2](https://user-images.githubusercontent.com/52317137/87081652-5acbe580-c247-11ea-98ff-36b35be3b822.PNG)

**Comment by avitial** [comment]
@CosmicHunter may I ask which version of OpenVINO are you using to convert to IR and which plugin are you using for inference? If you can also please provide ONNX model file(s), a sample input and the expected output.  Also I can see you are using IEPlugin class which is deprecated, try using [IECore](https://docs.openvinotoolkit.org/latest/ie_python_api/classie__api_1_1IECore.html).

**Comment by CosmicHunter** [comment]
I am using 2020.3.194 version of openvino toolkit .  the model is working fine with onnx runtime.
The command used for converting the model to onnx format
~~~
dummy_input = torch.autograd.Variable(torch.ones(1,200,dtype=torch.long))
dummy_h = torch.autograd.Variable(torch.zeros(2,1,256,dtype =torch.float32))
dummy_h = (dummy_h,dummy_h)
dummy_h
input_names = [ "input1", "hidden"]
output_names = [ "output" ]
net = net.float()
torch.onnx.export(net ,(dummy_input ,dummy_h),"lstm_model_2_test.onnx",input_names = input_names ,output_names = output_names,verbose=True,export_params=True)
~~~

The output when the model was converted to onnx gave something like this
![Screenshot_2020-07-10 Google Colaboratory](https://user-images.githubusercontent.com/52317137/87133259-b08aa700-c2b4-11ea-97d8-a061abd86bc3.png)

![Screenshot_2020-07-10 Google Colaboratory(1)](https://user-images.githubusercontent.com/52317137/87133266-b2546a80-c2b4-11ea-9582-7fb88e60dafe.png)

## Sample Input For Knowing the shapes
~~~
input1 = torch.randn(1,200)
input2 = torch.zeros(2,1,256)
input3 = torch.zeros(2,1,256)

input1 = input1.to(device)
input2 = input2.to(device)   # input2 and inputs3 will be used for further inferencing to initialize the hidden states
input3 = input3.to(device)
~~~

I have mailed you the onnx file .

**Comment by avitial** [comment]
@CosmicHunter I've received the model files. I am able to run the code snippet with the Inference Engine and the model in IR format with minor changes. For every inference the` print(res['Squeeze_24'])` statement results in `[0.5443078] `or` [0.54430777]`. Similarly with FP16 precision, results are `[0.54429036]`, `[0.5442904]` or `[0.5442905]`. What do you expect **Squeeze_24** results to be?

During ONNX->IR conversion, the Model Optimizer throws `warning: Convert data type of Parameter "input1" to int32`. Have you tried changing `input1 `data type to int32?

**Comment by CosmicHunter** [comment]
Yes I am also able to run it , but it is a sentiment classification model , and for negative as well as positive labels it is producing the same output always ,

**Comment by CosmicHunter** [comment]
The same onnx model with the onnx-runtime is running perfectly fine

**Comment by CosmicHunter** [comment]
No I have not converted the datatype of input1 to int32 , Can you please tell me how to do that , and whether to do it in pytorch model or the onnx model . I can share that code with you if you can please tell me

**Comment by avitial** [comment]
@CosmicHunter got it, you'd probably want to do this change in the PyTorch* model. Feel free to share those files as before, I can help investigate.

**Comment by CosmicHunter** [comment]
Ok Thanks , I have emailed you the link

**Comment by CosmicHunter** [comment]
I have removed the warning of datatypes while generating IR , but still the answer is always same , irrespective of the input
while inferencing

**Comment by mvafin** [comment]
Hi @CosmicHunter, I was looking into your issue, but I get consistent results between onnxruntime and OpenVINO 2020.4. I noticed that you initialize `input1` as `torch.randn(1,200)`, but this input in the model acts as `indices` to the `Gather` node and should be cast to an integer type, which gives you a small range of values (mostly -1,0,1 with some higher numbers rarely), that is why the output is very similar. Presumably, you need to change how you initialize `input.1`.

**Comment by CosmicHunter** [comment]
That is a dummy input for converting it to onnx , While inferencing on the onnx model I am getting the same results as that of pytorch model , mainly while inferencing on the openvino I am getting same output for different inputs

**Comment by CosmicHunter** [comment]
This time I converted the pytorch to onnx model using a sentence as an input instead of normal random input. And then converted it to IR , Then ran inference on it the result was same for a positive as well as negative input

![out3](https://user-images.githubusercontent.com/52317137/87840634-5abe8c00-c8be-11ea-9922-985c648b137f.PNG)

The answer is coming out to be same value for different input, I have also removed all warnings while onnx to IR conversion

**Comment by mvafin** [comment]
> ## Code Snippet Used For Inferencing
> The models forward function expects a sequence of length of 200 and hidden state and cell state of dimension (2,1,256) where is 2 is number of layers and 1 is the batch size and 256 is the hidden dimension
> 
> ```
>     plugin = IEPlugin(target_device)  
>     args = build_argparser().parse_args()
>     model_xml = args.model
>     model_bin = os.path.splitext(model_xml)[0] + ".bin"
>     log.info("Loading network files:\n\t{}\n\t{}".format(model_xml, model_bin))
>     device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
>     num_requests = 1
>     net = IENetwork(model=model_xml, weights=model_bin)
>     input_blob1 = next(iter(net.inputs))
>     input_blob2 = next(iter(net.inputs))
>     input_blob3 = next(iter(net.inputs))
>     hiden_init_state = torch.zeros(2,1,256)
>     hiden_init_state = hiden_init_state.to(device)     
>     # inputs = args.input[0]   #for getting input sentence  
>     count = 0
>     result = device_model.infer({input_blob1: inputs,
>                                  input_blob2 : hiden_init_state,
>                                  input_blob3: hiden_init_state})
>     
>     request_id = 0
>     status = device_model.requests[request_id].wait(-1)
>     
>     if status == 0:
>         res = device_model.requests[request_id].outputs
>     print(res['Squeeze_24'])
>     print(res.keys())
> ```

After looking into that code snippet I think the problem might be in how you getting the result of inferencing. The provided code snippet only outputs one result, could you provide the full code snippet for multiple results, that are always similar?

Some notes on the provided snippet:

- I don't see where `device_model` is initialized, it should be obtained by loading `net` into `plugin`, have you missed a line of code?
- How you get `inputs` value?
- You are running `infer` method of `device_model` which gives you `result`, where you can already get the result of your infer request by just printing the value `result['Squeeze_24']`. Instead of that, you are getting the result by using `requests`, which is another mechanism of getting the inference results, but you always have one infer request and you running inference in sync mode as far as I can see.
- You are using deprecated API `IENetwork` and `IEPluging`, try to use `IECore` instead.

**Comment by CosmicHunter** [comment]
I have emailed you the code
